mozilla-services/socorro
# bug 1626698: raw crash schema

This implements a raw crash schema just like the processed crash schema we did in bug 1764395. In addition to implementing the schema, this adds tests, a validator script, reworks the RawCrash API to use the schema, reworks the report view to use the schema, and adds some additional tests to the processed crash schema to assert invariants about source annotations.

Some other changes that are riding along with this:

1. removing support for `BIOS_Manufacturer` and `MemoryErrorCorrection` annotations. These were removed from Firefox in bug 1710152, so we no longer need to support them in crash ingestion.
2. removing support for `NumberOfProcessors` annotation. Looks like this hasn't been used in a long time and it's not in `CrashAnnotations.yaml`. The data is in the minidump now.
3. removing support for `AdapterRendererIDs` annotation. This hasn't been emitted by products in a long time and it's not in `CrashAnnotations.yaml`.

To verify this:

1. I went through the existing unit tests, fixed them up to cover the changes (and in some cases, fix errors in the tests that were unearthed by these changes), implemented new ones to check new invariants.
4. Downloaded data using the RawCrash API and no api token (thus, this is public data) from production and localhost and compared the output to make sure we weren't leaking data.
5. Downloaded data using the RawCrash API and an api token (thus, this is all the raw crash data--public and protected) from production and localhost and compared the output to make sure the raw crash api didn't change.
6. Went through the report view to make sure it continued to work and didn't break because of a change in permissions including the Crash Annotations and Raw Data and Minidumps tabs.
7. Used Firefox private window to go through crash reports on localhost to make sure they worked correctly when logged out.

I'm fairly confident everything is good at this point.
